# Proyecto del Día - Analizador de Ventas

## Objetivo

Crear un programa en Python que analice un conjunto de datos de ventas de una tienda. El programa debe realizar varias operaciones de Data Science para proporcionar información valiosa sobre las ventas de la tienda.

# Consigna

1. **Lectura de datos:** Crea un DataFrame que contenga los datos provistos en el archivo `Datos_Ventas_Tienda.csv` provisto en esta lección. El archivo incluye información como fecha de venta, categoría de producto, cantidad vendida y precio.

2. **Fusión de datos:** Crea un segundo DataFrame que contenga los datos del archivo `Datos_Ventas_Tienda2.csv` (también provisto) y concaténalos para tener un solo DataFrame con toda la información.

3. **Tratamiento de datos:** Utiliza Pandas para manipular estos datos. Deberás realizar tareas como limpieza de datos, filtrado y transformaciones básicas.

4. **Análisis de ventas:** Realiza análisis para responder preguntas como:

    - ¿Cuál es el producto más vendido?
    - ¿Cuál es el mes con más ventas?

5. **Datos agrupados:** Agrupa los datos por categoría de producto y analiza las ventas por categoría.

6. **Guardar resultados:** Al final, guarda el DataFrame completo (incluyendo la columna de meses) en un archivo CSV en tu ordenador.

## 1. Lectura de datos

In [1]:
import pandas as pd

In [2]:
datos_ventas_tienda_1_path: str = './data/Datos_Ventas_Tienda.csv'
datos_ventas_tienda_1: pd.DataFrame = pd.read_csv(datos_ventas_tienda_1_path)
datos_ventas_tienda_1

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
0    1/17/2023  Electrónic         7              200         1400
1     9/1/2023  Electrónic         8              200         1600
2    7/29/2023    Juguetes         3               30           90
3    10/4/2023   Alimentos         4               10           40
4    2/28/2023  Electrónic         6              200         1200
..         ...         ...       ...              ...          ...
995  4/27/2023   Alimentos         2               10           20
996   6/9/2023        Ropa        10               50          500
997  8/21/2023   Alimentos         8               10           80
998  3/28/2023  Electrónic         6              200         1200
999   3/7/2023  Electrónic         4              200          800

[1000 rows x 5 columns]

### Obtener información preliminar del DataFrame

In [3]:
datos_ventas_tienda_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Fecha            1000 non-null   object
 1   Producto         1000 non-null   object
 2   Cantidad         1000 non-null   int64 
 3   Precio Unitario  1000 non-null   int64 
 4   Total Venta      1000 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


## 2. Fusión de datos

In [4]:
datos_ventas_tienda_2_path: str = './data/Datos_Ventas_Tienda2.csv'
datos_ventas_tienda_2: pd.DataFrame = pd.read_csv(datos_ventas_tienda_2_path)
datos_ventas_tienda_2

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
0    8/25/2023  Electrónic         5              200         1000
1    8/14/2023   Alimentos         9               10           90
2    8/30/2023        Ropa         6               50          300
3    11/9/2023   Alimentos         2               10           20
4   10/26/2023        Ropa         8               50          400
5    12/1/2023    Juguetes         2               30           60
6    5/16/2023  Electrónic         4              200          800
7     2/9/2023      Libros         1               20           20
8     1/9/2023  Electrónic         2              200          400
9     5/3/2023  Electrónic         6              200         1200
10  11/30/2023  Electrónic         5              200         1000
11  10/26/2023  Electrónic         1              200          200
12   8/22/2023    Juguetes         3               30           90
13    6/5/2023  Electrónic         1              200          200
14   10/2/2023    Juguetes         6               30          180
15   1/25/2023  Electrónic         3              200          600
16   5/20/2023        Ropa         4               50          200
17   4/29/2023   Alimentos         8               10           80
18  12/20/2023        Ropa         2               50          100
19   4/10/2023        Ropa         8               50          400
20   5/19/2023    Juguetes         5               30          150
21    2/3/2023  Electrónic         4              200          800
22    3/4/2023        Ropa         5               50          250
23  11/12/2023    Juguetes         7               30          210
24   5/22/2023      Libros         7               20          140
25  11/19/2023        Ropa         4               50          200
26   7/31/2023   Alimentos         6               10           60
27  10/11/2023        Ropa         6               50          300
28   12/9/2023    Juguetes         4               30          120
29    3/6/2023      Libros         9               20          180
30   3/27/2023   Alimentos         1               10           10
31   4/23/2023        Ropa         9               50          450
32   5/26/2023  Electrónic         6              200         1200
33   1/20/2023    Juguetes         5               30          150
34   3/22/2023   Alimentos         9               10           90
35   2/26/2023      Libros         9               20          180
36   7/30/2023    Juguetes         1               30           30
37  11/18/2023        Ropa         3               50          150
38   6/20/2023      Libros         8               20          160
39    6/6/2023      Libros         3               20           60
40    5/2/2023   Alimentos         5               10           50
41    8/7/2023  Electrónic         9              200         1800
42  11/18/2023    Juguetes         8               30          240
43   1/17/2023    Juguetes         8               30          240
44  10/10/2023      Libros         6               20          120
45   9/11/2023    Juguetes        10               30          300
46    1/3/2023    Juguetes        10               30          300
47   11/7/2023   Alimentos         3               10           30
48   8/20/2023  Electrónic         6              200         1200
49   7/21/2023    Juguetes         3               30           90

### Obtener información preliminar del DataFrame

In [5]:
datos_ventas_tienda_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Fecha            50 non-null     object
 1   Producto         50 non-null     object
 2   Cantidad         50 non-null     int64 
 3   Precio Unitario  50 non-null     int64 
 4   Total Venta      50 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 2.1+ KB


### Concatenar los dos DataFrames

In [6]:
concat_df: pd.DataFrame = pd.concat(
    [datos_ventas_tienda_1, datos_ventas_tienda_2],
    axis=0
)
concat_df

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
0   1/17/2023  Electrónic         7              200         1400
1    9/1/2023  Electrónic         8              200         1600
2   7/29/2023    Juguetes         3               30           90
3   10/4/2023   Alimentos         4               10           40
4   2/28/2023  Electrónic         6              200         1200
..        ...         ...       ...              ...          ...
45  9/11/2023    Juguetes        10               30          300
46   1/3/2023    Juguetes        10               30          300
47  11/7/2023   Alimentos         3               10           30
48  8/20/2023  Electrónic         6              200         1200
49  7/21/2023    Juguetes         3               30           90

[1050 rows x 5 columns]

## 3. Tratamiento de datos

### Calcular los registros nulos

In [9]:
print(f'Cantidad de nulos por columna:\n{concat_df.isnull().sum()}')

Cantidad de nulos por columna:
Fecha              0
Producto           0
Cantidad           0
Precio Unitario    0
Total Venta        0
dtype: int64


### Chequear si hay duplicados

In [10]:
print(f'Cantidad de duplicados: {concat_df.duplicated().sum()}')

Cantidad de duplicados: 29


### Ver cuáles son los registros duplicados

In [11]:
duplicated: pd.DataFrame = concat_df[concat_df.duplicated(keep=False)]
duplicated

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
4     2/28/2023  Electrónic         6              200         1200
15    12/5/2023      Libros         4               20           80
43    4/11/2023   Alimentos         6               10           60
56   11/11/2023   Alimentos         1               10           10
57    9/28/2023      Libros         8               20          160
106   7/30/2023    Juguetes         1               30           30
124  10/26/2023        Ropa         3               50          150
129  10/12/2023    Juguetes         1               30           30
173    1/7/2023      Libros         2               20           40
188   6/29/2023   Alimentos        10               10          100
209    4/3/2023   Alimentos         1               10           10
214    1/6/2023      Libros         1               20           20
261  10/26/2023        Ropa         3               50          150
268    1/9/2023        Ropa         9               50          450
269   9/18/2023        Ropa        10               50          500
291   5/17/2023   Alimentos         1               10           10
301  12/14/2023    Juguetes         6               30          180
319   4/11/2023   Alimentos         6               10           60
326    1/6/2023      Libros         1               20           20
361   6/29/2023   Alimentos        10               10          100
362   4/29/2023   Alimentos         8               10           80
363  11/11/2023   Alimentos         1               10           10
383  10/12/2023    Juguetes         1               30           30
386   9/11/2023  Electrónic         7              200         1400
408    1/9/2023        Ropa         9               50          450
454   9/18/2023        Ropa        10               50          500
461   6/30/2023    Juguetes         3               30           90
492  12/14/2023    Juguetes         6               30          180
552   2/28/2023  Electrónic         6              200         1200
565   10/8/2023        Ropa         3               50          150
588    1/7/2023      Libros         2               20           40
596   2/19/2023      Libros         3               20           60
611   1/27/2023  Electrónic         4              200          800
638   12/5/2023      Libros         4               20           80
639   4/20/2023    Juguetes         3               30           90
643   10/8/2023        Ropa         3               50          150
656   7/19/2023        Ropa         1               50           50
673   2/28/2023      Libros         1               20           20
693    5/8/2023   Alimentos         4               10           40
699   4/11/2023    Juguetes         4               30          120
701    5/8/2023   Alimentos         4               10           40
716   8/21/2023   Alimentos         8               10           80
797   4/11/2023    Juguetes         4               30          120
801   9/11/2023  Electrónic         7              200         1400
826    4/3/2023   Alimentos         1               10           10
831   2/19/2023      Libros         3               20           60
839   5/17/2023   Alimentos         1               10           10
855   6/30/2023    Juguetes         3               30           90
865   8/14/2023        Ropa         4               50          200
909   2/28/2023      Libros         1               20           20
934   7/19/2023        Ropa         1               50           50
936   9/28/2023      Libros         8               20          160
944   4/20/2023    Juguetes         3               30           90
946   8/14/2023        Ropa         4               50          200
966   1/27/2023  Electrónic         4              200          800
997   8/21/2023   Alimentos         8               10           80
17    4/29/2023   Alimentos         8               10           80
36    7/30/2023    Juguetes         1               30           3

### Eliminar los registros duplicados

In [12]:
drop_duplicates: pd.DataFrame = concat_df.drop_duplicates()
drop_duplicates

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
0   1/17/2023  Electrónic         7              200         1400
1    9/1/2023  Electrónic         8              200         1600
2   7/29/2023    Juguetes         3               30           90
3   10/4/2023   Alimentos         4               10           40
4   2/28/2023  Electrónic         6              200         1200
..        ...         ...       ...              ...          ...
45  9/11/2023    Juguetes        10               30          300
46   1/3/2023    Juguetes        10               30          300
47  11/7/2023   Alimentos         3               10           30
48  8/20/2023  Electrónic         6              200         1200
49  7/21/2023    Juguetes         3               30           90

[1021 rows x 5 columns]

### Ver info de `drop_duplicates`

In [13]:
drop_duplicates.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1021 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Fecha            1021 non-null   object
 1   Producto         1021 non-null   object
 2   Cantidad         1021 non-null   int64 
 3   Precio Unitario  1021 non-null   int64 
 4   Total Venta      1021 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 47.9+ KB


### Resetear los índices

In [14]:
df_concat_reset: pd.DataFrame = drop_duplicates.reset_index(drop=True)
df_concat_reset

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
0     1/17/2023  Electrónic         7              200         1400
1      9/1/2023  Electrónic         8              200         1600
2     7/29/2023    Juguetes         3               30           90
3     10/4/2023   Alimentos         4               10           40
4     2/28/2023  Electrónic         6              200         1200
...         ...         ...       ...              ...          ...
1016  9/11/2023    Juguetes        10               30          300
1017   1/3/2023    Juguetes        10               30          300
1018  11/7/2023   Alimentos         3               10           30
1019  8/20/2023  Electrónic         6              200         1200
1020  7/21/2023    Juguetes         3               30           90

[1021 rows x 5 columns]

### Verificar el tipo de dato de un elemento de la columna 'Fecha'

In [15]:
print(type(df_concat_reset['Fecha'][0]))

<class 'str'>


### Cambiar el tipo de dato de la columna 'Fecha' al tipo correcto

In [16]:
df_concat_reset['Fecha'] = pd.to_datetime(
    df_concat_reset['Fecha'],
    format='%m/%d/%Y'
)
print(df_concat_reset['Fecha'][0])
print(type(df_concat_reset['Fecha'][0]))

2023-01-17 00:00:00
<class 'pandas._libs.tslibs.timestamps.Timestamp'>


### Reordenar las filas por la columna 'Fecha'

In [17]:
df_concat_reset.sort_values(by='Fecha', inplace=True)
df_concat_reset

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
369 2023-01-01   Alimentos         3               10           30
911 2023-01-01   Alimentos         2               10           20
703 2023-01-01      Libros         8               20          160
222 2023-01-02      Libros         6               20          120
97  2023-01-02  Electrónic         5              200         1000
..         ...         ...       ...              ...          ...
683 2023-12-30        Ropa         7               50          350
801 2023-12-30  Electrónic         4              200          800
419 2023-12-30  Electrónic         3              200          600
539 2023-12-31  Electrónic         8              200         1600
876 2023-12-31      Libros         8               20          160

[1021 rows x 5 columns]

### Volver a resetear los índices, según el nuevo orden

In [18]:
df_concat_reset = df_concat_reset.reset_index(drop=True)
df_concat_reset

Fecha    Producto  Cantidad  Precio Unitario  Total Venta
0    2023-01-01   Alimentos         3               10           30
1    2023-01-01   Alimentos         2               10           20
2    2023-01-01      Libros         8               20          160
3    2023-01-02      Libros         6               20          120
4    2023-01-02  Electrónic         5              200         1000
...         ...         ...       ...              ...          ...
1016 2023-12-30        Ropa         7               50          350
1017 2023-12-30  Electrónic         4              200          800
1018 2023-12-30  Electrónic         3              200          600
1019 2023-12-31  Electrónic         8              200         1600
1020 2023-12-31      Libros         8               20          160

[1021 rows x 5 columns]

## 4. Análisis de ventas

### 4.1. Obtener el producto más vendido (en general), contando la cantidad de productos vendidos por tipo.

In [19]:
grouped_by_product: pd.Series = df_concat_reset.groupby('Producto')['Cantidad'].sum()
grouped_by_product.sort_values(ascending=False)

Producto
Alimentos     1223
Electrónic    1203
Ropa          1098
Juguetes      1095
Libros        1023
Name: Cantidad, dtype: int64

**Respuesta:** Los productos de tipo "Alimentos" han sido los más vendidos según los registros, con un total de 1223 unidades vendidas. Adicionalmente, respondiendo a la pregunta contraria, la menor cantidad de ventas se registra en la categoría "Libros", con 1023.

### 4.2. Obtener el mes con más ventas en general, sin importar el tipo de producto que se haya vendido.

In [20]:
grouped_by_month: pd.Series = df_concat_reset.groupby(
    df_concat_reset['Fecha'].dt.to_period('M')
    )['Cantidad'].sum()
grouped_by_month.sort_values(ascending=False)

Fecha
2023-08    657
2023-01    548
2023-05    528
2023-03    483
2023-12    480
2023-10    479
2023-04    454
2023-06    447
2023-07    432
2023-09    416
2023-11    377
2023-02    341
Freq: M, Name: Cantidad, dtype: int64

#### Mejorando legibilidad de los resultados

In [21]:
month_mapping: dict[int, str] = {
    1: 'Enero', 2: 'Febrero', 3: 'Marzo', 4: 'Abril', 5: 'Mayo', 6: 'Junio',
    7: 'Julio', 8: 'Agosto', 9: 'Septiembre', 10: 'Octubre', 11: 'Noviembre',
    12: 'Diciembre'
}

grouped_by_month.index = grouped_by_month.index.map(
    lambda x: month_mapping[x.month] + ' ' + str(x.year)
)
grouped_by_month.sort_values(ascending=False)

Fecha
Agosto 2023        657
Enero 2023         548
Mayo 2023          528
Marzo 2023         483
Diciembre 2023     480
Octubre 2023       479
Abril 2023         454
Junio 2023         447
Julio 2023         432
Septiembre 2023    416
Noviembre 2023     377
Febrero 2023       341
Name: Cantidad, dtype: int64

**Respuesta:** El mes que registró más ventas en general fue Agosto, con 657. Por contraparte, el peor mes fue Febrero, con 341.

## 5. Datos agrupados

Agrupar los datos por **categoría de producto** y analizar las ventas por categoría.

**NOTA:** Parte de esto ya se hizo anteriormente en el apartado "Análisis de ventas", pero podemos ampliarlo para ver las ganancias.

In [22]:
profit_by_product: pd.Series = (
    df_concat_reset.groupby('Producto')['Total Venta'].sum()
)
profit_by_product.sort_values(ascending=False)

Producto
Electrónic    240600
Ropa           54900
Juguetes       32850
Libros         20460
Alimentos      12230
Name: Total Venta, dtype: int64

**Respuesta:** Las mayores ganancias del año provienen de la venta de productos de la categoría "Electrónica", con $240.600 en total. Irónicamente, la menor cantidad de ganancias proviene de la categoría "Alimentos", pese a que, en el análisis previo, se arrojó que la mayor cantidad de ventas fue por productos de dicha categoría.

## 6. Guardar resultados

In [23]:
# Guardar resultados en un archivo CSV
save_path: str = './data/Proyecto_Datos_Ventas.csv'
df_concat_reset.to_csv(save_path, index=False)
print(
    'Se ha guardado el archivo con el DataFrame completo en la ruta: '
    + save_path
)

Se ha guardado el archivo con el DataFrame completo en la ruta: ./data/Proyecto_Datos_Ventas.csv
